extensions/handler/message_handler.go

```diff
@@ -0,0 +1,176 @@
+package handler
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/sirupsen/logrus"
+	pushErrors "github.com/topfreegames/pusher/errors"
+	"github.com/topfreegames/pusher/extensions"
+	"github.com/topfreegames/pusher/interfaces"
+	"sync"
+	"time"
+)
+
+type messageHandler struct {
+	app               string
+	logger            *logrus.Logger
+	client            interfaces.PushClient
+	config            messageHandlerConfig
+	stats             messagesStats
+	statsMutex        sync.Mutex
+	feedbackReporters []interfaces.FeedbackReporter
+	statsReporters    []interfaces.StatsReporter
+}
+
+var _ interfaces.MessageHandler = &messageHandler{}
+
+func NewMessageHandler(
+	app string,
+	client interfaces.PushClient,
+	feedbackReporters []interfaces.FeedbackReporter,
+	statsReporters []interfaces.StatsReporter,
+	logger *logrus.Logger,
+) interfaces.MessageHandler {
+	l := logger.WithFields(logrus.Fields{
+		"app":    app,
+		"source": "messageHandler",
+	})
+	return &messageHandler{
+		app:               app,
+		client:            client,
+		feedbackReporters: feedbackReporters,
+		statsReporters:    statsReporters,
+		logger:            l.Logger,
+		config:            newDefaultMessageHandlerConfig(),
+	}
+}
+
+func (h *messageHandler) HandleMessages(ctx context.Context, msg interfaces.KafkaMessage) {
+	l := h.logger.WithFields(logrus.Fields{
+		"method": "HandleMessages",
+	})
+	km := extensions.KafkaGCMMessage{}
+	err := json.Unmarshal(msg.Value, &km)
+	if err != nil {
+		l.WithError(err).Error("Error unmarshaling message.")
+		return
+	}
+
+	if km.PushExpiry > 0 && km.PushExpiry < extensions.MakeTimestamp() {
+		l.Warnf("ignoring push message because it has expired: %s", km.Data)
+
+		h.statsMutex.Lock()
+		h.stats.ignored++
+		h.statsMutex.Unlock()
+
+		return
+	}
+
+	if km.Metadata != nil {
+		if km.Message.Data == nil {
+			km.Message.Data = map[string]interface{}{}
+		}
+
+		for k, v := range km.Metadata {
+			if km.Message.Data[k] == nil {
+				km.Message.Data[k] = v
+			}
+		}
+	}
+
+	err = h.client.SendPush(ctx, km.Message)
+	if err != nil {
+		l.WithError(err).Error("Error sending push message.")
+		h.statsMutex.Lock()
+		h.stats.failures++
+		h.statsMutex.Unlock()
```

Comment:
nit: consider using [Atomic.AddInt64](https://pkg.go.dev/sync/atomic#AddInt64) instead of Mutex just to lock and add 1 to this value